### ISSUE TITLE
[css-page] Should named pages be ignored when inside out-of-flow elements?

### ISSUE THREAD

**Issue by mstensho** [w3c/csswg-drafts#9513]
I cannot find anything in the spec suggesting that named pages are to be ignored inside floats or absolutely positioned elements. Yet, Firefox and PrinceXML don't support it. It is supported in Chromium, which makes the WPT test css/css-page/page-name-abspos-002-print.html fail in Chromium.

```html
<div style="position:absolute;">
  Page break after this line?
  <div style="page:b;">This should be on the next page, no?</div>
</div>
```

Spec: https://drafts.csswg.org/css-page-3/#using-named-pages

**Comment by mstensho** [comment]
This issue is about more than out-of-flow. 

external/wpt/css/css-page/page-name-flex-001-print.html fails similarly because it asserts that named pages on flex items should be ignored.

Again, Firefox and PrinceXML agree on ignoring them, whereas Chromium honors the page names.

**Comment by faceless2** [comment]
That's an interesting test. We don't currently honour it either, but I think it would be useful - it is essentially becomes a  "full page float", pushing an item to a page on its own like you might do for (for example) a color page in a monochrome book.

That mental model makes sense for out-of-flow; I'm not sure it does for things like flex-items, table-cells etc.

**Comment by mstensho** [comment]
The spec simply says that 'page' applies to boxes that create class A break points. In my mind that's pretty much aligned with where 'break-before' and 'break-after' apply, albeit with some exceptions, perhaps.

For column flex in particular - https://www.w3.org/TR/css-flexbox-1/#pagination: "In a column flex container, Class A break opportunities occur between sibling flex items".

Not sure about row flex containers. 'break-before' and 'break-after' properties are propagated to the flex line. One could imagine the same to happen for 'page', but then again maybe not so useful, since there may be multiple flex items specifying different page names for the same line.

**Comment by mstensho** [comment]
@dholbert Are these tests really valid, or should they be changed?

**Comment by dholbert** [comment]
> The spec simply says that 'page' applies to boxes that create class A break points. In my mind that's pretty much aligned with where 'break-before' and 'break-after' apply, albeit with some exceptions, perhaps.
> 
> For column flex in particular - https://www.w3.org/TR/css-flexbox-1/#pagination: "In a column flex container, Class A break opportunities occur between sibling flex items".

Thanks - that does make [`page-name-flex-001-print.html`](https://searchfox.org/firefox-main/rev/33fd6bd39c625067a29f153adce6a4646e45750f/testing/web-platform/tests/css/css-page/page-name-flex-001-print.html) invalid, then.  (Historical note, that was added in https://bugzilla.mozilla.org/show_bug.cgi?id=1786399 and I believe at that point we were going off of https://www.w3.org/TR/css-break-3/#btw-blocks as defining the complete list of things that can create class A break points -- we didn't notice that the flexbox spec added another case.)

> Not sure about row flex containers. 'break-before' and 'break-after' properties are propagated to the flex line. One could imagine the same to happen for 'page', but then again maybe not so useful, since there may be multiple flex items specifying different page names for the same line.

Yeah, not sure about those either; it's less obvious how the propagation (of page-names from item to line) would work there.

**Comment by dholbert** [comment]
> I cannot find anything in the spec suggesting that named pages are to be ignored inside floats or absolutely positioned elements.

It's a bit ambiguous.  As noted in my previous comment, we were looking at https://www.w3.org/TR/css-break-3/#btw-blocks as the main definition for class A break points (i.e. as the points where an adjacent "end page value" & "start page value" get compared), and that mentions these two cases, relating to in-flow boxes and floats:
```
Between sibling boxes of the following types:  [...]
* in-flow block-level boxes
* a float and an immediately-adjacent in-flow or floated box
```
We took the "in-flow" qualifier there as excluding out-of-flow stuff (and we took the special casing for floats as allowing breaks before/after floats but not within them).

However, I suppose there's a case to be made that e.g. **block-level siblings inside of an OOF** are technically "in-flow" within their local formatting context, and hence would be eligible for the above-quoted spec text to apply.  *shrug*

Its not clear to me whether the spec meant to include or exclude those.

**Comment by mstensho** [comment]
Yeah, in general, regular explicit breaks *inside* out-of-flow boxes should work, of course, but named pages has a rather unique piece of propagation speccery which seems rather oblivious to the concept of parallel fragmentation flows, and I think this is what requires clarification.
